include/pybind11/buffer_info.h

```diff
@@ -150,6 +153,11 @@ struct buffer_info {
     Py_buffer *view() const { return m_view; }
     Py_buffer *&view() { return m_view; }
 
+    template <typename T>
+    static bool compare(const buffer_info &b) {
```

Comment:
Can this be named has_same_format or something like that? compare sounds like a general == check, when this is only checking the format of the buffer.

Reply:
Yes, I totally agree, but what is a good name?

Here is my line of though that got me to the new name and source code comment:

"format" has so so many meanings already, to the uninitiated it likely means nothing. Additionally, the `Py_buffer` `format` strings may actually not be the same, because of the i/q/l ambiguity I was struggling with here.

Trying to start from the top down, what would a good source code comment be?

```
// Determines if the buffer item type is `T`.
```

Asking myself: Is that 100% accurate? What is "is" when it comes to aliases, such as `long double` being effectively an alias for `double` on some platforms? Or `long` effectively being an alias for `int` on some platforms, and `long long` an alias for `long` on others? (I think I actually stumbled over all of these ambiguities at various points in time far back.)

```
// Determines if the buffer item type is equivalent to `T`.
```

Asking myself for completeness: Do we have to worry about endianness? Searching for "endia", "most", "least", "signi" under https://docs.python.org/3/c-api/buffer.html doesn't pull up anything, which I take it means that we don't have to worry. It appears to be assumed that endianness conventions are never mixed.

Reply:
I think the precise semantics here are the same as https://en.cppreference.com/w/cpp/types/is_same

// Determines if the butter item type is the same as `T` (see std::is_same)

> Asking myself for completeness: Do we have to worry about endianness? Searching for "endia", "most", "least", "signi" under https://docs.python.org/3/c-api/buffer.html doesn't pull up anything, which I take it means that we don't have to worry. It appears to be assumed that endianness conventions are never mixed.

Good point. Yeah, your code is actually technically incorrect as you aren't handling endianness. But it will only provide false negatives. The format strings do specify endianness. This is specified in https://docs.python.org/3/library/struct.html#byte-order-size-and-alignment

Note that if there is no endian specifier, the default is "native", so your code won't return any false positives.

Your current code will return false negatives. "@i" and "=i" for example are equivalent to "int", but your code will return false. "<i" will be equivalent to int on little-endian platforms. ">i" and "!i" on big-endian platforms. Etc, etc.

Not sure if this is worth handling.

Reply:
> I think the precise semantics here are the same as https://en.cppreference.com/w/cpp/types/is_same

Not sure: last night I saw that `std::is_same<double, long double>::value` with MSVC is `false`, which is why I had to change my code to `sizeof(long double) == sizeof(double)` (b09e75b25c91bd300a1f2dbaabb1056a58d1d942), because that's what we really want to know.

> Not sure if this is worth handling.

I'm pretty sure no: https://docs.python.org/3/c-api/buffer.html isn't explicit about it (unfortunately), but it's strongly implied that endianness is assumed not to be mixed. Therefore I think going into that in the context of `buffer_info` will not help anything, on the contrary.

The situation is very different for serializing/deserializing arrays, then it has to be assumed endianness could be different between the producer and consumer, which could run on totally different platforms.

But the buffer interface is all about in-memory access. Hopefully we won't have one process (one address space) that mixes endianness. (It sounds like an idea from hell.)

Reply:
> Also, why is this a static method and not a member function?

Then you can take a simple function pointer, vs a member function pointer. I think that's easier to work with in generic code.

Other than that, it doesn't really matter?

Reply:
> Other than that, it doesn't really matter?

It's rather uncommon to see a static member function that could be a member function since member functions have a better interface for general use. I actually can't think of a single example in the standard library where a static member function is used like this.

It's more of a style thing though, not major.

Reply:
Ok, I'll try to change it to a member function asap. (I figure it's now or never, i.e. important to get right now.)

Reply:
Done: ef34d293495844933b1d936a058c5030c31f3ad3

I agree the `buffer_info` API looks better that way. The only trouble was/is, the syntax for dereferencing member function pointers badly rubs me the wrong way. `(buffer.request().*((*equiv_table)[cpp_name]))()`, what a beauty!